### PR TITLE
Add the support of lowering torch.cat to relay::concatenate

### PR DESCRIPTION
**Pull request by llyfacebook**

### CODE REVIEW

File: test/test_operators.py [hunk 1]
```diff
@@ -284,6 +284,24 @@ def linear_no_bias(input, weight):
         ref_out_no_bias, tvm_out_no_bias = self.runBoth(linear_no_bias, input, weight)
         assert torch.allclose(ref_out_no_bias, tvm_out_no_bias, rtol=0.01, atol=0.01)
 
+    @TVMTest.given(
+        shape=TVMTest.rand_shape(rank=2, min_dim=4),
+    )
+    def test_concat(self, shape):
+        input1 = torch.rand(shape)
+        input2 = torch.rand(shape)
+        # import pdb
```
Comment (kimishpatel) [inline test/test_operators.py]: Remove this.

File: test/test_operators.py [hunk 2]
```diff
@@ -284,6 +284,24 @@ def linear_no_bias(input, weight):
...
+        # pdb.set_trace()
+        def concat(x1, x2):
+            return torch.cat((x1, x2), 0)
+
+        import torch_tvm
+        torch_tvm.enable()
+        # test concat
+        scripted_concat = torch.jit.script(concat)
+        concat_graph = scripted_concat.graph_for(input1, input2)
+        FileCheck().check("prim::FusedConcat").check_not("prim::ListConstruct").check_not("aten::cat").run(str(concat_graph))
```
Comment (kimishpatel) [inline test/test_operators.py]: This seems to be the test not for concat but whether ListConstruct and cat are fuse. We should name it such.

File: test/test_operators.py [hunk 3]
```diff
@@ -284,6 +284,24 @@ def linear_no_bias(input, weight):
...
+        input2 = torch.rand(shape)
+        # import pdb
+        # pdb.set_trace()
+        def concat(x1, x2):
+            return torch.cat((x1, x2), 0)
+
+        import torch_tvm
+        torch_tvm.enable()
+        # test concat
+        scripted_concat = torch.jit.script(concat)
```
Comment (kimishpatel) [inline test/test_operators.py]: Also add a test for actual concat? You can use the same runBoth functionality for that.

File: torch_tvm/fuse_concat.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+using namespace torch::jit;
+
+bool isFusableCatNode(Node* node) {
+  if (node->kind() != aten::cat)
+    return false;
+  if (!node->is_constant(attr::dim))
+    return false;
+
+  auto tensors_node = node->namedInput(attr::tensors)->node();
+  if ((tensors_node->inputs().size() + node->outputs().size()) >
```
Comment (kimishpatel) [inline torch_tvm/fuse_concat.cpp]: Can you add a comment as to what is this supposed to check? And maybe `constexpr int some_var_name = 128`?

File: torch_tvm/fuse_concat.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  AT_ASSERT(node->kind() == aten::cat);
+  Graph* graph = node->owningGraph();
+  Node* list_construct = node->namedInput(attr::tensors)->node();
+  int64_t dim = node->get<int64_t>(attr::dim).value();
+
+  Node* fused_cat = graph->create(prim::FusedConcat, list_construct->inputs())
+                        ->i_(attr::dim, dim);
+  fused_cat->insertBefore(list_construct);
+  fused_cat->output()->copyMetadata(node->output());
+  node->output()->replaceAllUsesWith(fused_cat->output());
```
Comment (kimishpatel) [inline torch_tvm/fuse_concat.cpp]: Dont you want to check `node->output()->uses().empty()` and erase that as well?

File: torch_tvm/fuse_concat.cpp [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  if (list_construct->output()->uses().empty()) {
```
Comment (kimishpatel) [inline torch_tvm/fuse_concat.cpp]: Don't you have to first check this condition on node, destroy that and check it again on list_construct? Or does pytorch remove all the nodes without uses on its own?

File: torch_tvm/fuse_concat.cpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+void fuseConcats(Block* block_, AliasDb& aliasDb) {
+  for (auto it = block_->nodes().rbegin(); it != block_->nodes().rend();
+       ++it) {
+    Node* cat = *it;
+    if (!isFusableCatNode(cat)) {
+      continue;
+    }
+
+    Node* list_construct = cat->namedInput(attr::tensors)->node();
```
Comment (kimishpatel) [inline torch_tvm/fuse_concat.cpp]: Does not seem like `list_construct` is being used.

File: torch_tvm/fuse_concat.cpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    createFusedConcat(cat);
+  }
+}
+
+void FuseConcat(std::shared_ptr<Graph> graph) {
```
Comment (kimishpatel) [inline torch_tvm/fuse_concat.cpp]: @bwasti, should we construct aliasdb once and run all the fusion passes?

File: torch_tvm/fuse_concat.cpp [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  fused_cat->insertBefore(list_construct);
+  fused_cat->output()->copyMetadata(node->output());
+  node->output()->replaceAllUsesWith(fused_cat->output());
+  if (list_construct->output()->uses().empty()) {
+    list_construct->destroy();
+  }
+  return fused_cat;
+}
+
+void fuseConcats(Block* block_, AliasDb& aliasDb) {
```
Comment (kimishpatel) [inline torch_tvm/fuse_concat.cpp]: It doesnt seem that aliasDb is being used. Remove it.

File: torch_tvm/fuse_concat.cpp [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+      continue;
+    }
+
+    Node* list_construct = cat->namedInput(attr::tensors)->node();
+    createFusedConcat(cat);
+  }
+}
+
+void FuseConcat(std::shared_ptr<Graph> graph) {
+  AliasDb aliasDb(graph);
```
Comment (kimishpatel) [inline torch_tvm/fuse_concat.cpp]: Since it is not used this can be removed too.

Comment (kimishpatel) [reply torch_tvm/fuse_concat.cpp]: @bwasti, should we add unused-variable as error flag?

File: torch_tvm/fuse_concat.cpp (hunk 8 above)
Comment (yinghai) [reply torch_tvm/fuse_concat.cpp]: This is fine. It's done once.

File: torch_tvm/fuse_concat.cpp [hunk 11]
```diff
@@ -0,0 +1,59 @@
+#include "fusion_pass.h"
+#include "operators.h"
+
+using namespace torch::jit;
+
+bool isFusableCatNode(Node* node) {
+  if (node->kind() != aten::cat)
```
Comment (yinghai) [inline torch_tvm/fuse_concat.cpp]: Please add `{}` around your block.

File: torch_tvm/fuse_concat.cpp (hunk 8 above)
Comment (kimishpatel) [reply torch_tvm/fuse_concat.cpp]: I also suppose, once the graph is altered by earlier fusions aliasing information will change. So makes sense to do it for each fusion pass.

**Comment by yinghai** [comment]
@llyfacebook Let's land this?

File: test/test_operators.py (hunk 3 above)
Comment (llyfacebook) [reply test/test_operators.py]: sure

File: torch_tvm/fuse_concat.cpp (hunk 10 above)
Comment (llyfacebook) [reply torch_tvm/fuse_concat.cpp]: I extracted some code from pytorch side and didn't clean it up. :)

File: torch_tvm/fuse_concat.cpp (hunk 6 above)
Comment (llyfacebook) [reply torch_tvm/fuse_concat.cpp]: I think the replaceAllUsesWith do it internally.

File: torch_tvm/fuse_concat.cpp [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    list_construct->destroy();
+  }
+  return fused_cat;
+}
+
+void fuseConcats(Block* block_) {
+  for (auto it = block_->nodes().rbegin(); it != block_->nodes().rend();
+       ++it) {
+    Node* cat = *it;
+    if (!isFusableCatNode(cat)) {
```
Comment (kimishpatel) [inline torch_tvm/fuse_concat.cpp]: nit. Can be just `!isFusableCatNode(*it)`.

**Pull request merged**